oozie-el-extensions/src/main/java/org/apache/oozie/extensions/OozieELExtensions.java

```diff
@@ -72,15 +77,31 @@ public static String ph3_dataIn(String dataInName, String part) {
         //optional input
         if (uristr == null) {
             Element dsEle = getDSElement(eval, dataInName);
-            Configuration conf = new Configuration();
             SyncCoordAction appInst = (SyncCoordAction) eval.getVariable(CoordELFunctions.COORD_ACTION);
+
+            Configuration conf = new Configuration();
+            conf.set(OozieClient.USER_NAME, (String)eval.getVariable(OozieClient.USER_NAME));
             try {
                 ELEvaluator instEval = CoordELEvaluator.createInstancesELEvaluator(dsEle, appInst, conf);
                 StringBuilder instances = new StringBuilder();
-                CoordCommandUtils.resolveInstanceRange(dsEle, instances , appInst, conf, instEval);
-                uristr = CoordCommandUtils.createEarlyURIs(dsEle, instances.toString(),
-                        new StringBuilder(), new StringBuilder());
-                uristr = uristr.replace(CoordELFunctions.INSTANCE_SEPARATOR, ",");
+                StringBuilder urisWithDoneFlag = new StringBuilder();
+                CoordCommandUtils.resolveInstanceRange(dsEle, instances, appInst, conf, instEval);
+                CoordCommandUtils.createEarlyURIs(dsEle, instances.toString(),
+                        new StringBuilder(), urisWithDoneFlag);
+                XLog.getLog(OozieELExtensions.class).debug("Resolved instances for " + dataInName + " : "
+                        + urisWithDoneFlag.toString());
+                // Check if availability flags are present for each instance.
+                ActionDependency actionDep = DependencyChecker.checkForAvailability(urisWithDoneFlag.toString(),
+                        conf, false);
+                String doneFlag = (String) eval.getVariable(dataInName + ".done-flag");
+                uristr = StringUtils.join(stripDoneFlag(actionDep.getAvailableDependencies(), doneFlag), ",");
+                // If no instances are present, point the optional input to empty dir.
+                if (StringUtils.isEmpty(uristr)) {
+                    String emptyDir = (String) eval.getVariable(dataInName + ".empty-dir");
```

Comment:
Is this empty directory  different for different feeds ? Or we are pointing common empty directory which was created in Staging directory ?